### ISSUE TITLE
Generic types in a generic class

### ISSUE THREAD

**Issue by SweetPotato0213** [nus-cs2030/2021-s2#81]
# Summary

I am confused about all the different generic types in Box<>.

## Description 1:

What does the question mark mean in the expression `Box<?>`? Is it that all the generic class can store a null Box<Object> instance?

## Code Snippets

`private static final Box<?> EMPTY_BOX = new Box<Object>(null);`

## Description 2.1:

Why there is a `<T>` in front of `Box<T>`? What does the first `<T>` means? 

## Code Snippets

`static <T> Box<T> of(T t) {
        return t == null ? empty() : new Box<T>(t);
    }`

## Description 2.2:

But here, we don't have <T> in front of Box<T> as the `static <T> Box<T> of(T t)` method before.

## Code Snippets

`Box<T> filter(Predicate<? super T> predicate) {
        return predicate.test(object) ? this : empty();
    }`

## Description 3:

Also, the same issue, why there is a `<U>` in front of `Box<U>`?
For `function` in the `map` method, `? super T` is the input, does it mean that any function that works on the super type of T also works on T?
The output is `? extends U`. Does it mean that Box<U> can store any sub type of U?

## Code Snippets

`<U> Box<U> map(Function<? super T, ? extends U> function) {
        return isPresent() ? Box.of(function.apply(object)) : empty();
    }`

**Comment by Keithlimhs** [comment]
Description 1: The ? expression is a wildcard for generics. It means that EMPTY_BOX can be any type of Box generic.

Description 2.1: The <T> in front of Box< T > is the return type of the static method. The reason why there is a need to put it is because static methods are called from the Class itself, which in this case is a generic class which has not been given a type yet, unlike an instance method which has a generic type declared when instantiating the instance. Hence there is a need to include a <T> after the static keyword to indicate what the return type of the static method is, so that there is more type assurance. 

Description 2.2: There is no <T> in front of Box< T > in this case because the filter method is an instance method, not a static method, and it returns the Box< T > instance that the method was called from

**Comment by SweetPotato0213** [comment]
> Description 1: The ? expression is a wildcard for generics. It means that EMPTY_BOX can be any type of Box generic.
> 
> Description 2.1: The in front of Box< T > is the return type of the static method. The reason why there is a need to put it is because static methods are called from the Class itself, which in this case is a generic class which has not been given a type yet, unlike an instance method which has a generic type declared when instantiating the instance. Hence there is a need to include a after the static keyword to indicate what the return type of the static method is, so that there is more type assurance.
> 
> Description 2.2: There is no in front of Box< T > in this case because the filter method is an instance method, not a static method, and it returns the Box< T > instance that the method was called from

Thank you for your reply! It really helps! 
I think now I can understand why in Description 3 there is a `<U>` in front of `Box<U>`! (Is it that the return type of the static method is `<U>` as it has not been given a type yet?)

**Comment by dumbasssoul** [comment]
iinw the <U> is the type parameter for static methods, which should come before the return type, which is actually Box<U> in this case

**Comment by gwajoon** [comment]
> > Description 1: The ? expression is a wildcard for generics. It means that EMPTY_BOX can be any type of Box generic.
> > Description 2.1: The in front of Box< T > is the return type of the static method. The reason why there is a need to put it is because static methods are called from the Class itself, which in this case is a generic class which has not been given a type yet, unlike an instance method which has a generic type declared when instantiating the instance. Hence there is a need to include a after the static keyword to indicate what the return type of the static method is, so that there is more type assurance.
> > Description 2.2: There is no in front of Box< T > in this case because the filter method is an instance method, not a static method, and it returns the Box< T > instance that the method was called from
> 
> Thank you for your reply! It really helps!
> I think now I can understand why in Description 3 there is a `<U>` in front of `Box<U>`! (Is it that the return type of the static method is `<U>` as it has not been given a type yet?)

An instance method can also define the generic type parameter (<T>) in front of the return type, but it doesn't have to, since it can use a generic type parameter already defined in the class level (public class Box<T>).

On the other hand, a static method cannot use the generic type parameter defined in the class level, so it must declare any generic type parameter it intends to use, hence you have to put '<U>' in front of the return type of the static method being called. Hope that clarifies the issue for you!

**Comment by SweetPotato0213** [comment]
> > > Description 1: The ? expression is a wildcard for generics. It means that EMPTY_BOX can be any type of Box generic.
> > > Description 2.1: The in front of Box< T > is the return type of the static method. The reason why there is a need to put it is because static methods are called from the Class itself, which in this case is a generic class which has not been given a type yet, unlike an instance method which has a generic type declared when instantiating the instance. Hence there is a need to include a after the static keyword to indicate what the return type of the static method is, so that there is more type assurance.
> > > Description 2.2: There is no in front of Box< T > in this case because the filter method is an instance method, not a static method, and it returns the Box< T > instance that the method was called from
> > 
> > 
> > Thank you for your reply! It really helps!
> > I think now I can understand why in Description 3 there is a `<U>` in front of `Box<U>`! (Is it that the return type of the static method is `<U>` as it has not been given a type yet?)
> 
> An instance method can also define the generic type parameter () in front of the return type, but it doesn't have to, since it can use a generic type parameter already defined in the class level (public class Box).
> 
> On the other hand, a static method cannot use the generic type parameter defined in the class level, so it must declare any generic type parameter it intends to use, hence you have to put '' in front of the return type of the static method being called. Hope that clarifies the issue for you!

Ah yes! I am clearer now. Thank you~

**Comment by SweetPotato0213** [comment]
> iinw the `<U>` is the type parameter for static methods, which should come before the return type, which is actually `Box<U>` in this case

But I think `map` is not a static function right?

**Comment by gableejh** [comment]
> > iinw the `<U>` is the type parameter for static methods, which should come before the return type, which is actually `Box<U>` in this case
> 
> But I think `map` is not a static function right?

map shouldn't be static because it's not shared or common between different students etc.

**Comment by SweetPotato0213** [comment]
> > > iinw the `<U>` is the type parameter for static methods, which should come before the return type, which is actually `Box<U>` in this case
> > 
> > 
> > But I think `map` is not a static function right?
> 
> map shouldn't be static because it's not shared or common between different students etc.

then why you say that <U> us the type parameter for static methods?

**Comment by gableejh** [comment]
> > > > iinw the `<U>` is the type parameter for static methods, which should come before the return type, which is actually `Box<U>` in this case
> > > 
> > > 
> > > But I think `map` is not a static function right?
> > 
> > 
> > map shouldn't be static because it's not shared or common between different students etc.
> 
> then why you say that us the type parameter for static methods?

Apologies, please ignore my previous comment - mixed it up with another question. But for this map method, it's not static. The reason why there is a <U> in front is because the type U is only defined within this method and only exists within this method as well. When we say defined within this method, it means that the type U is not defined in the class level, in this case it's type T which is defined. If you take a look at the Function, it gives out "? extends U", which basically means type U plus any of its subclass(es). 

To sum up, you should add a <X> (can be any alphabet) in front of the method for (1) static methods and (2) when a new type is returned by the function, such as Function<T,R> or BiFunction<U, T, U>. In this case, you should add a <R> in front of the method for Function and <U> in front of the method for BiFunction. 

Hope this helps

**Comment by SweetPotato0213** [comment]
> > > > > iinw the `<U>` is the type parameter for static methods, which should come before the return type, which is actually `Box<U>` in this case
> > > > 
> > > > 
> > > > But I think `map` is not a static function right?
> > > 
> > > 
> > > map shouldn't be static because it's not shared or common between different students etc.
> > 
> > 
> > then why you say that us the type parameter for static methods?
> 
> Apologies, please ignore my previous comment - mixed it up with another question. But for this map method, it's not static. The reason why there is a in front is because the type U is only defined within this method and only exists within this method as well. When we say defined within this method, it means that the type U is not defined in the class level, in this case it's type T which is defined. If you take a look at the Function, it gives out "? extends U", which basically means type U plus any of its subclass(es).
> 
> To sum up, you should add a (can be any alphabet) in front of the method for (1) static methods and (2) when a new type is returned by the function, such as Function<T,R> or BiFunction<U, T, U>. In this case, you should add a in front of the method for Function and in front of the method for BiFunction.
> 
> Hope this helps

Ah yes! It really helps! thank Thank you so much :)
